File: lib-database/src/main/java/com/ustadmobile/core/db/dao/DownloadJobItemDao.java
package com.ustadmobile.core.db.dao;

import com.ustadmobile.core.db.JobStatus;
import com.ustadmobile.core.db.UmAppDatabase;
import com.ustadmobile.core.db.UmLiveData;
import com.ustadmobile.core.impl.UmCallback;
import com.ustadmobile.lib.database.annotation.UmDao;
import com.ustadmobile.lib.database.annotation.UmInsert;
import com.ustadmobile.lib.database.annotation.UmQuery;
import com.ustadmobile.lib.database.annotation.UmTransaction;
import com.ustadmobile.lib.database.annotation.UmUpdate;
import com.ustadmobile.lib.db.entities.ConnectivityStatus;
import com.ustadmobile.lib.db.entities.ContainerWithContentEntry;
import com.ustadmobile.lib.db.entities.DownloadJob;
import com.ustadmobile.lib.db.entities.DownloadJobItem;
import com.ustadmobile.lib.db.entities.DownloadJobItemParentChildJoin;
import com.ustadmobile.lib.db.entities.DownloadJobItemStatus;
import com.ustadmobile.lib.db.entities.DownloadJobItemWithDownloadSetItem;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * DAO for the DownloadJobItem class
 */
@UmDao
public abstract class DownloadJobItemDao {

    public static class DownloadJobItemToBeCreated{

        long downloadSetItemUid;

        long containerUid;

        long fileSize;

        long contentEntryUid;

        public long getDownloadSetItemUid() {
            return downloadSetItemUid;
        }

        public void setDownloadSetItemUid(long downloadSetItemUid) {
            this.downloadSetItemUid = downloadSetItemUid;
        }

        public long getContainerUid() {
            return containerUid;
        }

        public void setContainerUid(long containerUid) {
            this.containerUid = containerUid;
        }

        public long getFileSize() {
            return fileSize;
        }

        public void setFileSize(long fileSize) {
            this.fileSize = fileSize;
        }

        public long getContentEntryUid() {
            return contentEntryUid;
        }

        public void setContentEntryUid(long contentEntryUid) {
            this.contentEntryUid = contentEntryUid;
        }
    }

    public static class DownloadJobItemToBeCreated2 {

        long cepcjUid;

        long contentEntryUid;

        long containerUid;

        long fileSize;

        long parentEntryUid;

        public long getContentEntryUid() {
            return contentEntryUid;
        }

        public void setContentEntryUid(long contentEntryUid) {
            this.contentEntryUid = contentEntryUid;
        }

        public long getContainerUid() {
            return containerUid;
        }

        public void setContainerUid(long containerUid) {
            this.containerUid = containerUid;
        }

        public long getFileSize() {
            return fileSize;
        }

        public void setFileSize(long fileSize) {
            this.fileSize = fileSize;
        }

        public long getParentEntryUid() {
            return parentEntryUid;
        }

        public void setParentEntryUid(long parentEntryUid) {
            this.parentEntryUid = parentEntryUid;
        }

        public long getCepcjUid() {
            return cepcjUid;
        }

        public void setCepcjUid(long cepcjUid) {
            this.cepcjUid = cepcjUid;
        }
    }


    /**
     * Insert a list of DownloadJobItems
     *
     * @param jobRunItems List of DownloadJobItem to insert
     */
    @UmInsert
    public abstract void insert(List<DownloadJobItem> jobRunItems);

    @UmTransaction
    public void insertListAndSetIds(List<DownloadJobItem> jobItems) {
        for(DownloadJobItem item : jobItems) {
            item.setDjiUid(insert(item));
        }
    }

    @UmTransaction
    public void updateDownloadJobItemsProgress(List<DownloadJobItemStatus> statusList) {
        for(DownloadJobItemStatus status : statusList) {
           updateDownloadJobItemProgress(status.getJobItemUid(), status.getBytesSoFar(),
                   status.getTotalBytes());
        }
    }

    @UmQuery("UPDATE DownloadJobItem SET downloadedSoFar = :bytesSoFar, " +
            "downloadLength = :totalLength WHERE djiUid = :djiUid")
    public abstract void updateDownloadJobItemProgress(int djiUid, long bytesSoFar, long totalLength);



    @UmQuery("SELECT DownloadJobItem.*, DownloadSetItem.* FROM DownloadJobItem " +
            "LEFT JOIN DownloadSetItem ON DownloadJobItem.djiDsiUid = DownloadSetItem.dsiUid " +
            "WHERE DownloadJobItem.djiUid = :djiUid")
    public abstract DownloadJobItemWithDownloadSetItem findByUidWithDownloadSetItem(long djiUid);

    @UmUpdate
    public abstract int update(DownloadJob downloadJob);

    @UmUpdate
    public abstract void updateList(List<DownloadJobItem> downloadJobList);

    /**
     * Insert a single DownloadJobItem
     *
     * @param jobRunItem DownloadJobItem to insert
     */
    @UmInsert
    public abstract long insert(DownloadJobItem jobRunItem);

    @UmQuery("DELETE FROM DownloadJobItem")
    public abstract void deleteAll(UmCallback<Void> callback);

    /**
     * Update the main status fields for the given DownloadJobitem
     *
     * @param djiUid DownloadJobItemId to updateState (primary key)
     * @param djiStatus status property to set
     * @param downloadedSoFar downloadedSoFar property to set
     * @param downloadLength downloadLength property to set
     * @param currentSpeed currentSpeed property to set
     */
    @UmQuery("Update DownloadJobItem SET " +
            "djiStatus = :djiStatus, downloadedSoFar = :downloadedSoFar, " +
            "downloadLength = :downloadLength, currentSpeed = :currentSpeed " +
            " WHERE djiUid = :djiUid")
    protected abstract void updateDownloadJobItemStatusIm(long djiUid, int djiStatus,
                                                     long downloadedSoFar, long downloadLength,
                                                     long currentSpeed);

    public  void updateDownloadJobItemStatus(long djiUid, int djiStatus,
                                                     long downloadedSoFar, long downloadLength,
                                                     long currentSpeed) {
        System.out.println("updateDownloadJobItemStatus " + djiUid + " -> " + djiStatus);
        updateDownloadJobItemStatusIm(djiUid, djiStatus, downloadedSoFar, downloadLength,
                currentSpeed);
    }

    @UmQuery("UPDATE DownloadJobItem SET downloadedSoFar = :downloadedSoFar, " +
            "currentSpeed = :currentSpeed " +
            "WHERE djiUid = :djiUid")
    public abstract void updateDownloadJobItemProgress(long djiUid, long downloadedSoFar,
                                                       long currentSpeed);


    @UmQuery("UPDATE DownloadJobItem SET djiStatus = :status WHERE djiUid = :djiUid")
    protected abstract void updateItemStatusInt(long djiUid, long status);

    public void updateStatus(long djiUid, long status){
        System.out.println("DownloadJob #" +djiUid+ " updating status to " + status);
        updateItemStatusInt(djiUid,status);
    }


    @UmQuery("UPDATE DownloadJobItem SET numAttempts = numAttempts + 1 WHERE djiUid = :djiUid")
    public abstract void incrementNumAttempts(long djiUid);

    @UmQuery("SELECT DownloadJobItem.*, DownloadSetItem.* FROM " +
            "DownloadJobItem " +
            "LEFT JOIN DownloadSetItem ON DownloadJobItem.djiDsiUid = DownloadSetItem.dsiUid " +
            "WHERE DownloadJobItem.djiUid = :djiUid")
    public abstract DownloadJobItemWithDownloadSetItem findWithDownloadSetItemByUid(long djiUid);

    @UmQuery("SELECT djiStatus FROM DownloadJobItem WHERE djiUid = :djiUid")
    public abstract UmLiveData<Integer> getLiveStatus(long djiUid);

    @UmQuery("SELECT * FROM DownloadJobItem")
    public abstract UmLiveData<List<DownloadJobItem>> findAllLive();

    @UmQuery("SELECT * FROM DownloadJobItem")
    public abstract List<DownloadJobItem> findAll();

    @UmQuery("SELECT COUNT(*) FROM DownloadJobItem WHERE djiDjUid =:djiDjUid")
    public abstract int  getTotalDownloadJobItems(long djiDjUid);

    @UmQuery("SELECT COUNT(*) FROM DownloadJobItem WHERE djiDjUid =:djiDjUid")
    public abstract UmLiveData<Integer> countDownloadJobItems(long djiDjUid);

    @UmQuery("SELECT DownloadSetItem.dsiUid as downloadSetItemUid, Container.fileSize, \n" +
            "Container.containerContentEntryUid as contentEntryUid , Container.containerUid\n" +
            "FROM DownloadSetItem LEFT JOIN Container \n" +
            "ON Container.containerUid = (SELECT Container.containerUid FROM Container \n" +
            "WHERE containerContentEntryUid =  DownloadSetItem.dsiContentEntryUid ORDER BY lastModified  DESC LIMIT 1) \n" +
            "WHERE  Container.fileSize != 0 AND DownloadSetItem.dsiDsUid = :downloadSetUid")
    public abstract List<DownloadJobItemToBeCreated> findJobItemsToBeCreatedDownloadSet(long downloadSetUid);

    @UmQuery("SELECT destinationFile FROM DownloadJobItem WHERE djiUid != 0 AND djiDsiUid IN(:djiDsiUids)")
    public abstract List<String> getDestinationFiles(List<Long> djiDsiUids);


    @UmQuery("SELECT DownloadJobItem.*, DownloadSetItem.* FROM DownloadJobItem " +
            "LEFT JOIN DownloadSetItem ON DownloadJobItem.djiDsiUid = DownloadSetItem.dsiUid " +
            "LEFT JOIN DownloadSet on DownloadSetItem.dsiDsUid = DownloadSet.dsUid " +
            "WHERE " +
            " DownloadJobItem.djiContainerUid != 0 " +
            " AND DownloadJobItem.djiStatus >= " + JobStatus.WAITING_MIN +
            " AND DownloadJobItem.djiStatus < " + JobStatus.RUNNING_MIN +
            " AND (((SELECT connectivityState FROM ConnectivityStatus) =  " + ConnectivityStatus.STATE_UNMETERED + ") " +
                " OR ((SELECT connectivityState FROM ConnectivityStatus) = " + ConnectivityStatus.STATE_METERED + ") " +
                " AND DownloadSet.meteredNetworkAllowed) " +
            "LIMIT 1")
    public abstract UmLiveData<List<DownloadJobItemWithDownloadSetItem>> findNextDownloadJobItems();

    @UmQuery("SELECT DownloadJobItem.*, DownloadSetItem.* FROM DownloadJobItem " +
            "LEFT JOIN DownloadSetItem ON DownloadJobItem.djiDsiUid = DownloadSetItem.dsiUid " +
            "WHERE DownloadSetItem.dsiContentEntryUid = :contentEntryUid " +
            "ORDER BY DownloadJobItem.timeStarted DESC LIMIT 1")
    public abstract DownloadJobItemWithDownloadSetItem findByContentEntryUid(long contentEntryUid);


    @UmQuery("SELECT * " +
            "FROM DownloadJobItem " +
            "WHERE djiContentEntryUid = :contentEntryUid " +
            "ORDER BY DownloadJobItem.timeStarted DESC LIMIT 1")
    public abstract DownloadJobItem findByContentEntryUid2(long contentEntryUid);


    @UmQuery("SELECT * " +
            "FROM DownloadJobItem " +
            "WHERE djiContentEntryUid = :contentEntryUid " +
            "ORDER BY DownloadJobItem.timeStarted DESC LIMIT 1")
    public abstract UmLiveData<DownloadJobItem> findByContentEntryUidLive(long contentEntryUid);

    @UmQuery("SELECT DownloadJobItem.* " +
            "FROM DownloadJobItem " +
            "WHERE DownloadJobItem.djiUid IN (:contentEntryUids) " +
            "ORDER BY DownloadJobItem.timeStarted DESC LIMIT 1")
    public abstract List<DownloadJobItem> findByDjiUidsList(List<Long> contentEntryUids);

    @UmQuery("SELECT djiUid AS jobItemUid, " +
            "djiContentEntryUid AS contentEntryUid, " +
            "downloadedSoFar AS bytesSoFar, " +
            "downloadLength AS totalBytes," +
            "djiStatus AS state " +
            "FROM DownloadJobItem " +
            "WHERE DownloadJobItem.djiDjUid = :downloadJobUid")
    public abstract List<DownloadJobItemStatus> findStatusByDownlaodJobUid(long downloadJobUid);

    @UmQuery("SELECT ContentEntryParentChildJoin.cepcjUid AS cepcjUid, " +
            " ContentEntryParentChildJoin.cepcjChildContentEntryUid AS contentEntryUid," +
            " Container.containerUid AS containerUid," +
            " Container.fileSize AS fileSize," +
            " ContentEntryParentChildJoin.cepcjParentContentEntryUid AS parentEntryUid " +
            " FROM ContentEntryParentChildJoin " +
            " LEFT JOIN Container ON Container.containerContentEntryUid = " +
            "   ContentEntryParentChildJoin.cepcjChildContentEntryUid" +
            "   AND Container.lastModified = " +
            "   (SELECT MAX(lastModified) FROM Container WHERE containerContentEntryUid = ContentEntryParentChildJoin.cepcjChildContentEntryUid) " +
            "WHERE " +
            "ContentEntryParentChildJoin.cepcjParentContentEntryUid in (:parentContentEntryUids)")
    public abstract List<DownloadJobItemToBeCreated2> findByParentContentEntryUuids(
            List<Long> parentContentEntryUids);

    @UmInsert
    public abstract void insertDownloadJobItemParentChildJoin(DownloadJobItemParentChildJoin dj);

    @UmTransaction
    public void updateJobItemStatusList(List<DownloadJobItemStatus> statusList) {
        for(DownloadJobItemStatus status : statusList) {
            updateStatus(status.getJobItemUid(), status.getStatus());
        }
    }

}

File: lib-database/src/commonMain/kotlin/com/ustadmobile/core/db/dao/DownloadJobItemDao.kt
package com.ustadmobile.core.db.dao

import androidx.room.*
import com.ustadmobile.core.db.JobStatus
import com.ustadmobile.core.db.JobStatus.COMPLETE
import com.ustadmobile.door.DoorLiveData
import com.ustadmobile.door.annotation.QueryLiveTables
import com.ustadmobile.lib.db.entities.*

/**
 * DAO for the DownloadJobItem class
 */
@Dao
abstract class DownloadJobItemDao {

    class DownloadJobItemToBeCreated2 {

        var cepcjUid: Long = 0

        var contentEntryUid: Long = 0

        var containerUid: Long = 0

        var fileSize: Long = 0

        var parentEntryUid: Long = 0
    }


    @Transaction
    open fun insertListAndSetIds(jobItems: List<DownloadJobItem>) {
        for (item in jobItems) {
            item.djiUid = insert(item).toInt()
        }
    }

    @Transaction
    open fun updateDownloadJobItemsProgressList(statusList: List<DownloadJobItemStatus>) {
        for (status in statusList) {
            updateDownloadJobItemProgress(status.jobItemUid, status.bytesSoFar,
                    status.totalBytes)
        }
    }

    @Query("UPDATE DownloadJobItem SET downloadedSoFar = :bytesSoFar, " + "downloadLength = :totalLength WHERE djiUid = :djiUid")
    abstract fun updateDownloadJobItemProgress(djiUid: Int, bytesSoFar: Long, totalLength: Long)

    @Update
    abstract fun update(downloadJob: DownloadJob): Int

    @Update
    abstract fun updateList(downloadJobList: List<DownloadJobItem>)

    /**
     * Insert a single DownloadJobItem
     *
     * @param jobRunItem DownloadJobItem to insert
     */
    @Insert
    abstract fun insert(jobRunItem: DownloadJobItem): Long

    @Query("UPDATE DownloadJobItem SET djiStatus = :status WHERE djiUid = :djiUid")
    abstract fun updateItemStatusInt(djiUid: Int, status: Int)

    @Transaction
    open fun updateStatus(djiUid: Int, status: Int) {
        println("DownloadJob #$djiUid updating status to $status")
        updateItemStatusInt(djiUid, status)
    }


    @Query("SELECT DownloadJobItem.* FROM " +
            "DownloadJobItem " +
            "WHERE DownloadJobItem.djiUid = :djiUid")
    abstract fun findByUid(djiUid: Int): DownloadJobItem?

    @Query("""SELECT DownloadJobItem.* FROM
        DownloadJobItem WHERE djiDjUid = :djUid AND
        djiContentEntryUid = (SELECT djRootContentEntryUid FROM DownloadJob WHERE djUid = :djUid)""")
    abstract fun findRootForDownloadJob(djUid: Int): DownloadJobItem?


    @Query("SELECT djiStatus FROM DownloadJobItem WHERE djiUid = :djiUid")
    abstract fun getLiveStatus(djiUid: Int): DoorLiveData<Int>

    @Query("SELECT * FROM DownloadJobItem")
    abstract fun findAllLive(): DoorLiveData<List<DownloadJobItem>>

    @Query("SELECT * FROM DownloadJobItem")
    abstract fun findAll(): List<DownloadJobItem>

    @Query("SELECT COUNT(*) FROM DownloadJobItem WHERE djiDjUid =:djiDjUid")
    abstract fun getTotalDownloadJobItems(djiDjUid: Int): Int

    @Query("SELECT COUNT(*) FROM DownloadJobItem WHERE djiDjUid =:djiDjUid")
    abstract suspend fun getTotalDownloadJobItemsAsync(djiDjUid: Int): Int

    @Query("SELECT COUNT(*) FROM DownloadJobItem WHERE djiDjUid =:djiDjUid")
    abstract fun countDownloadJobItems(djiDjUid: Long): DoorLiveData<Int>

    @Query("SELECT destinationFile FROM DownloadJobItem WHERE djiUid != 0 AND djiDsiUid IN(:djiDsiUids)")
    abstract fun getDestinationFiles(djiDsiUids: List<Long>): List<String>

    @Query("""SELECT DownloadJobItem.* FROM DownloadJobItem 
            LEFT JOIN DownloadJob ON DownloadJobItem.djiDjUid = DownloadJob.djUid 
            WHERE 
             DownloadJobItem.djiContainerUid != 0 
             AND DownloadJobItem.djiStatus >= ${JobStatus.WAITING_MIN} 
             AND DownloadJobItem.djiStatus < ${JobStatus.RUNNING_MIN} 
             AND (((SELECT connectivityState FROM ConnectivityStatus) =  ${ConnectivityStatus.STATE_UNMETERED} ) 
             OR (((SELECT connectivityState FROM ConnectivityStatus) = ${ConnectivityStatus.STATE_METERED} ) 
                AND DownloadJob.meteredNetworkAllowed) 
             OR EXISTS(SELECT laContainerUid FROM LocallyAvailableContainer WHERE 
                laContainerUid = DownloadJobItem.djiContainerUid)) 
             ORDER BY DownloadJob.timeRequested, DownloadJobItem.djiUid LIMIT 6""")
    @QueryLiveTables(["DownloadJobItem", "ConnectivityStatus", "DownloadJob"])
    abstract fun findNextDownloadJobItems(): DoorLiveData<List<DownloadJobItem>>

    @Query("""SELECT DownloadJobItem.* FROM DownloadJobItem 
            LEFT JOIN DownloadJob ON DownloadJobItem.djiDjUid = DownloadJob.djUid 
            WHERE 
             DownloadJobItem.djiContainerUid != 0 
             AND DownloadJobItem.djiStatus >= ${JobStatus.WAITING_MIN} 
             AND DownloadJobItem.djiStatus < ${JobStatus.RUNNING_MIN} 
             AND (:unmeteredNetworkAvailable OR DownloadJob.meteredNetworkAllowed 
             OR EXISTS(SELECT laContainerUid FROM LocallyAvailableContainer WHERE 
                laContainerUid = DownloadJobItem.djiContainerUid))
            ORDER BY DownloadJob.timeRequested, DownloadJobItem.djiUid LIMIT :limit
    """)
    abstract fun findNextDownloadJobItems2(limit: Int, unmeteredNetworkAvailable: Boolean): List<DownloadJobItem>


    @Query("SELECT DownloadJobItem.* FROM DownloadJobItem " +
            "WHERE DownloadJobItem.djiContentEntryUid = :contentEntryUid " +
            "ORDER BY DownloadJobItem.djiUid DESC LIMIT 1")
    abstract fun findByContentEntryUid(contentEntryUid: Long): DownloadJobItem?


    @Query("SELECT * " +
            "FROM DownloadJobItem " +
            "WHERE djiContentEntryUid = :contentEntryUid " +
            "ORDER BY DownloadJobItem.timeStarted DESC LIMIT 1")
    abstract fun findByContentEntryUid2(contentEntryUid: Long): DownloadJobItem?


    @Query("SELECT * " +
            "FROM DownloadJobItem " +
            "WHERE djiContentEntryUid = :contentEntryUid " +
            "ORDER BY DownloadJobItem.timeStarted DESC LIMIT 1")
    abstract fun findByContentEntryUidLive(contentEntryUid: Long): DoorLiveData<DownloadJobItem?>

    @Query("""SELECT Container.containerUid, Container.mimeType from DownloadJobItem 
        LEFT JOIN Container ON DownloadJobItem.djiContainerUid = Container.containerUid 
        WHERE DownloadJobItem.djiContentEntryUid = :contentEntryUid AND DownloadJobItem.djiStatus = $COMPLETE
        ORDER BY Container.cntLastModified DESC LIMIT 1
    """)
    abstract suspend fun findMostRecentContainerDownloaded(contentEntryUid: Long): ContainerUidAndMimeType?

    @Query("SELECT DownloadJobItem.* " +
            "FROM DownloadJobItem " +
            "WHERE DownloadJobItem.djiUid IN (:contentEntryUids) " +
            "ORDER BY DownloadJobItem.timeStarted DESC LIMIT 1")
    abstract fun findByDjiUidsList(contentEntryUids: List<Long>): List<DownloadJobItem>

    @Query("SELECT djiUid AS jobItemUid, " +
            "djiContentEntryUid AS contentEntryUid, " +
            "downloadedSoFar AS bytesSoFar, " +
            "downloadLength AS totalBytes," +
            "djiStatus AS status " +
            "FROM DownloadJobItem " +
            "WHERE DownloadJobItem.djiDjUid = :downloadJobUid")
    abstract fun findStatusByDownlaodJobUid(downloadJobUid: Int): List<DownloadJobItemStatus>

    @Query("SELECT ContentEntryParentChildJoin.cepcjUid AS cepcjUid, " +
            " ContentEntryParentChildJoin.cepcjChildContentEntryUid AS contentEntryUid," +
            " Container.containerUid AS containerUid," +
            " Container.fileSize AS fileSize," +
            " ContentEntryParentChildJoin.cepcjParentContentEntryUid AS parentEntryUid " +
            " FROM ContentEntryParentChildJoin " +
            " LEFT JOIN Container ON Container.containerContentEntryUid = " +
            "   ContentEntryParentChildJoin.cepcjChildContentEntryUid" +
            "   AND Container.cntLastModified = " +
            "   (SELECT MAX(cntLastModified) FROM Container WHERE containerContentEntryUid = ContentEntryParentChildJoin.cepcjChildContentEntryUid) " +
            "WHERE " +
            "ContentEntryParentChildJoin.cepcjParentContentEntryUid in (:parentContentEntryUids)")
    abstract fun findByParentContentEntryUuids(
            parentContentEntryUids: List<Long>): List<DownloadJobItemToBeCreated2>

    @Query("""SELECT DownloadJobItem.* FROM DownloadJobItem 
        LEFT JOIN DownloadJobItemParentChildJoin ON DownloadJobItemParentChildJoin.djiChildDjiUid = DownloadJobItem.djiUid
        WHERE DownloadJobItemParentChildJoin.djiParentDjiUid IN (:parentDownloadJobUids)
         """)
    abstract fun findByParentDownloadJobUids(parentDownloadJobUids: List<Int>): List<DownloadJobItem>

    /**
     * Runs a given function block for each level of child download job items
     */
    fun forAllChildDownloadJobItemsRecursive(parentDownloadJobUid: Int, block: (batch: List<DownloadJobItem>) -> Unit) {
        var lastParentUids = listOf(parentDownloadJobUid)
        block.invoke(listOf(findByUid(parentDownloadJobUid)?: throw IllegalArgumentException("Please provide the parent content Entry")))
        do {
            val childItems = findByParentDownloadJobUids(lastParentUids)
            block.invoke(childItems)
            lastParentUids = childItems.filter { it.djiContainerUid == 0L }. map { it.djiUid }
        } while(childItems.isNotEmpty())
    }

    /**
     * Runs a given function block for each level of child download job items
     */
    suspend fun forAllChildDownloadJobItemsRecursiveAsync(parentDownloadJobUid: Int, block: suspend (batch: List<DownloadJobItem>) -> Unit) {
        var lastParentUids = listOf(parentDownloadJobUid)
        block.invoke(listOf(findByUid(parentDownloadJobUid)?: throw IllegalArgumentException("Please provide the parent content Entry")))
        do {
            val childItems = findByParentDownloadJobUids(lastParentUids)
            block.invoke(childItems)
            lastParentUids = childItems.filter { it.djiContainerUid == 0L }. map { it.djiUid }
        } while(childItems.isNotEmpty())
    }


    @Insert
    abstract fun insertDownloadJobItemParentChildJoin(dj: DownloadJobItemParentChildJoin)

    @Transaction
    open fun updateJobItemStatusList(statusList: List<DownloadJobItemStatus>) {
        for (status in statusList) {
            updateStatus(status.jobItemUid, status.status)
        }
    }

    @Query("SELECT * FROM DownloadJobItem WHERE djiDjUid = :downloadJobUid")
    abstract fun findByDownloadJobUid(downloadJobUid: Int): List<DownloadJobItem>

    @Query("""UPDATE DownloadJobItem SET djiStatus = :status, 
        downloadedSoFar = :downloadedSoFar, 
        downloadLength = :downloadLength
        WHERE djiUid = :djiUid""")
    abstract fun updateStatusAndProgress(djiUid: Int, status: Int, downloadedSoFar: Long, downloadLength: Long)

    @Transaction
    open fun updateStatusAndProgressList(downloadJobItems: List<DownloadJobItem>) {
        downloadJobItems.forEach {
            updateStatusAndProgress(it.djiUid, it.djiStatus, it.downloadedSoFar, it.downloadLength)
        }
    }

    @Query("""SELECT djiUid, djiStatus FROM DownloadJobItem WHERE djiUid IN 
        (SELECT djiChildDjiUid FROM DownloadJobItemParentChildJoin WHERE djiParentDjiUid = :parentDjiUid)""")
    abstract fun getUidAndStatusByParentJobItem(parentDjiUid: Int): List<DownloadJobItemUidAndStatus>


    /**
     * Update the status of any waiting items.
     */
    @Query("""UPDATE DownloadJobItem SET djiStatus = :status 
            WHERE djiDjUid = :downloadJobId
            AND DownloadJobItem.djiStatus < ${JobStatus.RUNNING_MIN} 
            AND DownloadJobItem.djiUid NOT IN (:excludedJobItemUids)""")
    abstract fun updateWaitingItemStatus(downloadJobId: Int, status: Int, excludedJobItemUids: List<Int>)
}
